core/shared/src/main/scala/zio/concurrent/ConcurrentSet.scala

```diff
@@ -11,15 +12,47 @@ final class ConcurrentSet[A] private (private val underlying: ConcurrentHashMap.
     UIO(underlying.add(x))
 
   def addAll(xs: Iterable[A]): UIO[Boolean] =
-    UIO {
-      var added = false
-      xs.foreach(x => added = underlying.add(x))
-      added
-    }
+    UIO(underlying.addAll(xs.asJavaCollection))
+
+  def collectFirst[B](pf: PartialFunction[A, B]): UIO[Option[B]] =
+    UIO(underlying.asScala.toSet.collectFirst(pf))
+
+  def exists(p: A => Boolean): UIO[Boolean] =
+    UIO(underlying.asScala.toSet.exists(p))
+
+  def fold[R, E, S](zero: S)(f: (S, A) => S): UIO[S] =
+    UIO(underlying.asScala.toSet.foldLeft(zero)(f))
+
+  def foldZIO[R, E, S](zero: S)(f: (S, A) => ZIO[R, E, S]): ZIO[R, E, S] =
+    ZIO.foldLeft(underlying.asScala.toSet)(zero)(f)
+
+  def foreach[R, E, B](f: A => ZIO[R, E, B]): ZIO[R, E, Set[B]] =
+    ZIO.foreach(underlying.asScala.toSet)(f)
```

Comment:
Here if we want to provide this we don't have other choice than to pass use the RTS. I'd be inclined to follow Adam's suggestion that we don't add effectual functions.